### PR TITLE
Post-ingest SNS topic

### PR DESCRIPTION
**Pull request by jkeifer**
I have been working on a solution for #45 in an older fork of this codebase. This PR is my attempt to port those changes to the new codebase. The basic ideas are tested in that fork, though it is old enough not to have an index per collection.

Note that I do not have a test environment setup for this newer version so this initial pass is as-of-yet completely untested. Any help with that would be appreciated. :)

It also looks like the automated ingest tests have been neglected and need some attention, which would probably help here.

Anyway, this set of changes is to allow stac-server ingest to post a message to an SNS topic for every item it processes, with some indication of whether ingest was successful or not (which is current swallowed up and the only inkling of an issue is a single log message). The goal is to allow services that need to respond to newly-available items a way to subscribe to a feed of those items.

Subscriptions to the post-ingest SNS topic can use filters to choose which messages the want to see, which include `ingestStatus` (`successful` or `failed`), `collection`, and `recordType` (like `Feature` for items or `Collection` for collections). I am open to additional message attributes to add to this set if any seem critical to include at this time.
Note that a lot of the changes here are related to removing the existing concept of streaming records to the ElasticSearch writer. The reasons for that change are two-fold:

* With the streams, there was no (obvious) way to pass up a failed promise to the outer context sending the data into the stream, or to allow access to any outer non-stream state (which was problematic for handling the notification record content and ensuring the messages would be as useful as possible).
* The streams didn't seem to provide any real benefit with regard to efficiency or scaling. First of all, the number of items processed by an ingest execution is already limited by the default SQS lambda trigger batch size. Second, the existing stream model more or less ensured multiple ES queries were required when ingesting more than one item per execution. With the new model all input items are batched into a single ES query, reducing round trips and increasing efficiency. I believe the new implementation will actually be more performant when large numbers of items need to be ingested.

### CODE REVIEW

**Comment by jkeifer** [comment]
The recent changes to the tests found some additional issues. I am stuck on the following ES errors:

```
Error ingesting collection2_item: {"type":"mapper_exception","reason":"the [enabled] parameter can't be updated for the object mapping [assets]"}
Error ingesting LC80100102015050LGN00: {"type":"mapper_exception","reason":"the [enabled] parameter can't be updated for the object mapping [assets]"}
Error ingesting LC80100102015082LGN00: {"type":"mapper_exception","reason":"the [enabled] parameter can't be updated for the object mapping [assets]"}
```

I saw similar errors when trying to ingest collections before converting the collections to a one-by-one ingest approach rather than using the bulk API (for other reasons as well, like ensuring the collections exist before trying to add the items to them). So it seems like these errors are related to the bulk API. Maybe it only is a problem for the first ingested item in a collection? When running the bulk API elsewhere with ES that already has items I don't see this issue.

Maybe someone with more ES experience than me would have some insight?

**Comment by jkeifer** [comment]
The issue was caused because the `type` parameter had been removed from transformed ES `record`s, but was still referenced in the bulk operations so I didn't realize the hardcoded `doc` type shouldn't have been included.

**Comment by matthewhanson** [comment]
@marchuffnagle can you review this change from using streams?

**Comment by marchuffnagle** [comment]
Can you please add tests for this new functionality?

**Review by marchuffnagle — CHANGES_REQUESTED** [review]
Please add tests

**Comment by marchuffnagle** [comment]
I'd suggest waiting until #179 is merged, that's going to include a number of utilities that will be useful in this PR.

**Comment by jkeifer** [comment]
@marchuffnagle Thanks for taking a look, I appreciate it. #179 definitely looks like it answers a lot of the questions I had about tests, so I'll wait for that to get in. It seems like you did most of the hard work there, which is great because I am _way_ over time working on this feature. :)

**Comment by marchuffnagle** [comment]
@jkeifer #179 is now merged. Please merge the main branch into your PR and I'll go through and review it.

**Comment by jkeifer** [comment]
@marchuffnagle Well it took me a lot of work to stay up-to-date with all the changes you have been getting in, my goodness. :) Just once I thought I finally got all the tests passing I saw you had more test changes and everything broke again.

I had some issues with the `ts-check` and a function I had tried to add to the ingest system tests. For the sake of time I just moved it into `tests/helpers/ingest-sns.js`. I know that's probably not ideal, but it has been a _long_ day.

Anyway, the tests are passing now. Let me know if you have comments or questions.

File: src/lib/ingest.js
Comment (marchuffnagle) [inline src/lib/ingest.js:237]: Where is it being enforced that collections are ingested one-by-one? It looks like all of the collections are being ingested in parallel.

File: src/lib/ingest.js
Comment (marchuffnagle) [inline src/lib/ingest.js:229]: This implementation could make it easier to understand what's happening here, and in what order:

```js
async ingest(records) {
  const [collections, otherItems] = partition(records, stacUtils.isCollection)

  // Ingest collections first
  await Promise.allSettled(collections.map((c) => this.ingestCollection(c)))

  const [items, invalidItems] = partition(otherItems, stacUtils.isItem)

  // Execute error handlers for invalid items
  const invalidItemHandlers = invalidItems.map(
    (i) => this.errorHandler(i, 'Unable to determine record type')
  )

  // Wait for items to be ingested and error handlers to run
  await Promise.allSettled([
    this.ingestItems(items),
    ...invalidItemHandlers
  ])
}
```

File: src/lib/ingest.js
Comment (marchuffnagle) [inline src/lib/ingest.js:107]: Rather than trying to maintain this collections state, I think we'd be better off just inserting the item and catching an error if the index does not exist.

**Review by marchuffnagle — CHANGES_REQUESTED** [review]

**Comment by marchuffnagle** [comment]
I agree that the stream-based ingest approach should be evaluated in the future. I do have a few concerns about the approach being taken here, though.

1. We're conflating two things in this one PR - removing stream-based ingest and publishing to SNS
2. Issue #45 , which this resolves, is targeted at release 0.5.0. We're still trying to get 0.4.0 out the door, and I'm worried about a large change like this impacting the current release.
3. The test coverage on the project is improving, but I'm worried about a change of this size introducing regressions that we would miss.

If we want to add the "publish to SNS" functionality right now, I think we should look at a smaller PR.

One possibility wold be to update `src/lambdas/ingest/index.js` to use `ingestItem` rather than `ingestItems`. By doing that, you can the publish a per-item success or failure message to SNS. The ingest handler could be updated with something like:

```js
await Promise.all(stacItems.map(async (i) => {
  try {
    await ingest.ingestItem(i, stream)
    await publishSuccessMessage()
  } catch (error) {
    await publishFailureMessage()
  }
}))
```

That could provide the new functionality with less risk of introducing a regression or impacting the 0.4.0 release.

**Comment by jkeifer** [comment]
The proposed idea with using `ingestItem` was the first thing I tried for implementing the post-ingest SNS. However, I could not see any way to get the promise from the ElasticSearch update operation in the stream writer `_write` method up to that `ingestItem` context for handling. It also seems like that approach could have performance implications; without load testing it is hard to know if that would be problematic for things like bulk ingests (@matthewhanson expressed some concerns about this based on his experiences with large ingests).

To put this more explicitly, https://github.com/stac-utils/stac-server/blob/main/src/lib/esStream.js#L69 currently swallows any errors returned from the ES update request. Removing that catch results in any error being uncaught despite the outer handlers.

For bulk ingests (which currently do occur when multiple items are batched up by the SQS lambda trigger), the response isn't even logged per item, so issues with invalid item formats are completely silent.

I would consider these two issues to be one big bug that needs to be fixed. If this can't be added to the scope for 0.4 even though I've already worked on it then I don't know what to say. Sure, it is conflating the post-ingest SNS topic with resolving that problem, but unless I am missing something it seems to be a requirement to do so to support the SNS topic in a usable way. 🤷

File: src/lib/ingest.js
Comment (jkeifer) [reply src/lib/ingest.js:107]: That's fair, I was just trying to keep with the existing functionality.

File: src/lib/ingest.js
Comment (jkeifer) [reply src/lib/ingest.js:237]: When I say one-by-one I mean using individual requests to ES rather than a single bulk request like is used for items.

**Comment by marchuffnagle** [comment]
> To put this more explicitly, [https://github.com/stac-utils/stac-server/blob/main/src/lib/esStream.js#L69](https://github.com/stac-utils/stac-server/blob/main/src/lib/esStream.js?rgh-link-date=2022-02-24T19%3A28%3A17Z#L69) currently swallows any errors returned from the ES update request. Removing that catch results in any error being uncaught despite the outer handlers.

The catch block you're referring to is:

```js
} catch (err) {
  logger.error(err)
  next()
}
```

If that were changed to

```js
} catch (err) {
  next(err)
}
```

then I think it should propagate the error out of the `ElasticSearchWritableStream` instance.

**Comment by jkeifer** [comment]
Hmm, well, that's interesting to learn. Would you happen to know of any good documentation about that `next` function that includes details about the fact that it can take an argument? I am relatively new to node, and in particular streams, but I did a fair bit of searching around about this paradigm and I didn't find any example where an error was passed into `next`.

Anyway, it sounds like this change set isn't going to be accepted. I haven't really had time to spend on putting this together, nor trying to keep it current so it could be considering for review, so I really don't have time right now to try it all over again. I guess we can just close this for now.

**Comment by marchuffnagle** [comment]
> Would you happen to know of any good documentation about that `next` function that includes details about the fact that it can take an argument?

https://nodejs.org/docs/latest-v14.x/api/stream.html#stream_writable_write_chunk_encoding_callback_1

"The callback function must be called synchronously inside of writable._write() or asynchronously (i.e. different tick) to signal either that the write completed successfully or failed with an error. The first argument passed to the callback must be the Error object if the call failed or null if the write succeeded."

**Pull request closed without merging**
